solidity/contracts/AltheaDB.sol

```diff
@@ -1,54 +1,369 @@
+                get_registered_client_with_wg_key(entry.wg_key),
+                get_null_identity()
+            )
+        ) {
+            return true;
+        }
+        return false;
+    }
+
+    // start user and exit management functions
 
     // Add a new registered user
-    function add_registered_user(Identity memory entry) public {
-        registered_users.push(entry);
-        wg_key_to_reg_users_map[entry.wg_key] = entry;
-        mesh_ip_to_reg_users_map[entry.mesh_ip] = entry;
-        eth_addr_to_reg_users_map[entry.eth_addr] = entry;
+    function add_registered_user(Identity calldata entry) public {
+        if (is_user_admin(msg.sender)) {
+            // if any client or exit currently registered has overlapping data, do not allow the
+            // registration to continue
+            if (check_for_any_duplicates(entry)) {
+                revert DuplicateUser();
+            }
+
+            state_registeredUsers.push(entry);
+            emit UserRegisteredEvent(entry);
+        } else {
+            revert UnathorizedCaller();
+        }
+    }
+
+    // Remove a new registered user
+    function remove_registered_user(Identity calldata entry) public {
+        if (is_user_admin(msg.sender)) {
+            uint256 index = get_index_of_id(entry, state_registeredUsers);
+            delete_array_entry(index, state_registeredUsers);
+            emit UserRemovedEvent(entry);
+        } else {
+            revert UnathorizedCaller();
+        }
+    }
+
+    // Add a new registered exit
+    function add_registered_exit(Identity calldata entry) public {
+        if (is_exit_admin(msg.sender)) {
+            // if any client or exit currently registered has overlapping data, do not allow the
+            // registration to continue
+            if (check_for_any_duplicates(entry)) {
+                revert DuplicateUser();
+            }
+
+            state_registeredExits.push(entry);
+            emit ExitRegisteredEvent(entry);
+        } else {
+            revert UnathorizedCaller();
+        }
     }
 
-    // Get all registered users
-    function get_all_registered_users()
-        public
-        view
-        returns (Identity[] memory)
-    {
-        return registered_users;
+    // Remove a new registered exit
+    function remove_registered_exit(Identity calldata entry) public {
+        if (is_exit_admin(msg.sender)) {
+            uint256 index = get_index_of_id(entry, state_registeredExits);
+            delete_array_entry(index, state_registeredExits);
+            emit ExitRemovedEvent(entry);
+        } else {
+            revert UnathorizedCaller();
+        }
     }
 
+    // start user query functions
+
+
     function get_registered_client_with_wg_key(
-        string memory wg_key
+        uint256 wg_key
     ) public view returns (Identity memory) {
-        return wg_key_to_reg_users_map[wg_key];
+        for (uint256 i = 0; i < state_registeredUsers.length; i++) {
```

Comment:
Why was the mapping(uint256 => Identity) removed and doing iteration instead?

Reply:
EVM code optimization requires a bit of background.

1) storage is very expensive, in fact it's a lot more expensive than computation  in the EVM the cost of every operation is strictly defined for gas purposes 

https://ethereum.org/en/developers/docs/evm/opcodes/

As you can see here iteration, addition and comparison operations are in the low single digit like 3,5,8 gas. But a store to a new data slot can cost up to a whopping 20,000 gas. 

So simply put the iteration is cheaper (it's also cheaper to store the uint256 rather than strings, because, strings have a length value that takes up a storage slot of it's own, as well as unicode encoding being less efficient)

2) the mapping is just more data to handle when doing deletion.

Reply:
Thats pretty cool, thanks for the clarificaiton